notes/plutus-report/costing.tex

```diff
@@ -0,0 +1,339 @@
+This is not perfectly precise, but it is reasonably correlated with the measures of size that e.g. addition is likely to care about.
+
+\paragraph{Builtin accounting happens before execution}
+Builtin operations can in principle use a lot of resources just in the operation itself.
+For example, multiplying two sufficiently large numbers can require a large amount of memory to hold the result, much larger than that required to hold the arguments.
+This leads to a conflict: builtin operations are atomic (in that we can't interrupt them), but we want to interrupt evaluation as soon as we exceed our resource budget.
+
+Our solution is to compute the costs for a builtin operation before we run it.
+Since we are not actually measuring anything, but rather computing the cost from our model, we are able to do this.
+That way we can terminate evaluation if we would exceed the budget after running the operation, without actually running the operation.
+
+\paragraph{Time for builtin operations}
+The approach we take for Time is very empirical:
+\begin{enumerate}
+\item Micro-benchmark the builtin with arguments of varying sizes.
+\item Look at the resulting graphs and try to pick an appropriate linear model for the running time based on the argument sizes (e.g. ``linear in both arguments'').
+\item Use linear regression to infer the parameters for the model, these are the cost model parameters for this builtin.
+\end{enumerate}
+
+\paragraph{Space for builtin operations}
+The approach we take for Space is more deductive.
+Generally it is hard to measure memory allocation, especially when it occurs in a foreign library.
+So we largely rely on reasoning about or inspecting the libraries in question.
+Many perform no allocation, or a predictable amount of allocation.
+
+\subsubsection{Space and Time for machine steps}
+The CEK machine itself has many steps that it goes through, which correspond to handling e.g. application of user-defined functions.
+These also need to have costs associated: naive profiling suggests that the operation machine itself (as opposed to builtin functions) is usually >50\% of evaluation time.
+
+\paragraph{Time for machine steps}
+The approach we initially took for Time is empirical, matching the approach for builtins:
+\begin{enumerate}
+\item Assume that the execution time is linear in all the kinds of machine steps.
+\item Benchmark execution of a large number of varied programs that use all the execution steps.
+\item Use linear regression to infer the parameters for the model, these are the cost parameters for the machine steps.
+\end{enumerate}
+
+However, when we actually did this, we found that the number of machine steps of different kinds are almost perfectly correlated, such that there is little point trying to infer coefficients for them individually.
+So instead we fit a single-parameter linear model based only on the total number of steps.
```

Comment:
I'd argue that forcing a delay should be cheaper than applying a lambda, but I agree that we don't to obsess with these details right now, maybe in V2.

Reply:
Empirics say it's not noticeable :shrug:

Reply:
What that refers to is just the cost of evaluating the `Apply` clause in the CEK `compute` function.  The actual cost of evaluating the body of a `LamAbs`, for example (which could even be infinite), is accounted for when the individual nodes within it are evaluated.  The costs of evaluating the individual clauses of `compute` seem to be pretty much the same, since none of them do a great deal of work.  

Also, I've looked at the number of times the various AST nodes are evaluated in lots of examples, and the frequencies with which the different node types appear seem to be pretty much constant across different examples (see the first table [here](https://github.com/input-output-hk/plutus/blob/master/notes/plutus-core/Merklisation/AST-analysis.md) for example, although that's (a) for typed PLC, and (b) static as opposed to dynamic).  This means that If we know the total number of compute steps we can pretty much work out the number of times the different node types are processed without actually needing to count them, which in turn means that the statistical models can't distinguish between them.  Now maybe you could cook up a program where 90% of the nodes are `force` or `delay` and that program would violate what I've just said, but I think that that still wouldn't make a lot of difference.

One thing that does make things a little tricky is that accessing variables in environments takes a non-uniform amount of time, and that will grow as the environments grow (although I haven't in fact seen this making any difference).  Having constant-time access to environment entries would help with this, but maybe it doesn't really matter.

Reply:
> What that refers to is just the cost of evaluating the Apply clause in the CEK compute function. The actual cost of evaluating the body of a LamAbs, for example (which could even be infinite), is accounted for when the individual nodes within it are evaluated. The costs of evaluating the individual clauses of compute seem to be pretty much the same, since none of them do a great deal of work.

That's what I'm talking about, yes. Evaluating a `Delay` is a constant-time operation, while evaluating a `LamAbs` requires inserting something into the environment, which is not a constant-time operation. Not only that, but also each `LamAbs` makes subsequent `LamAbs` and `Var` cases more expensive. So I'm pretty sure that `LamAbs` is generally more expensive than `Delay`, but maybe in our fairly small programs the constant dominates the asymptotic.

> One thing that does make things a little tricky is that accessing variables in environments takes a non-uniform amount of time, and that will grow as the environments grow (although I haven't in fact seen this making any difference). Having constant-time access to environment entries would help with this, but maybe it doesn't really matter.

Yep.

Reply:
Note that even if `LamAbs` was predictably twice as expensive as `Delay`, it could *still* be fitted well by a single-parameter model if the number of `LamAbs`s and `Delay`s is very strongly correlated. And empirically that seems to be the case.

Reply:
(Oh, that's what I said in the text, I'm just repeating myself)